src/threads-version/big_int64.cpp

```
@@ -0,0 +1,767 @@
+  std::random_device rd;
+  std::mt19937_64 gen(rd());
+  std::uniform_int_distribution<uint64_t> distrib(min, max);
+  return distrib(gen);
+}
+void BigInt64::zero() {
+  limbs.clear();
+  limbs.push_back(0);
+  isNegative = false;
+}
+BigInt64 longDivision(const BigInt64 &a, const BigInt64 &b,
+                      BigInt64 &remainder) {
+  // c=a/b;
+  if (b.isZero())
+    throw std::invalid_argument("Invalid input: non-digit character");
+  remainder.isNegative = b.isNegative;
+  if (a.isSmallerThanUnsigned(b)) {
+    remainder.limbs = a.limbs;
+    return BigInt64();
+  }
+  remainder.limbs.clear();
+  std::vector<uint64_t> quotient;
+
+  for (int i = a.limbsCount() - 1; i >= 0; i--) { // msb-lsb
+
+    remainder.limbs.insert(remainder.limbs.begin(),
+                           a.limbs[i]); // insert msb as lsb
+    if (remainder.limbsCount() > 1 && remainder.limbs.back() == 0)
+      remainder.limbs.pop_back(); // remove leading zero
+    uint64_t times = 0;
+    if (!remainder.isSmallerThanUnsigned(b)) {
+      uint64_t left = 1, right = MAX_VAL_64;
+
+      // after inserting next limb, most times t contains b is max value
+      // of singe limb, find in range of [0...max]
+      while (left <= right) {
+        uint64_t mid = left + (right - left) / 2;
+        if (!remainder.isSmallerThanUnsigned(b * mid)) { // b*mid<=remainder
+          times = mid;
+          if (left == right)
+            break;
+          left = mid + 1;
+        } else {
+          right = mid - 1;
+        }
+      }
+      remainder.subtractUnsigned(b * times);
+    }
+
+    if (quotient.size() > 0 || times > 0)
+      quotient.push_back(times);
+  }
+
+  std::reverse(quotient.begin(), quotient.end());
+  BigInt64 quoti(a.isNegative != b.isNegative);
+  quoti.limbs = std::move(quotient);
+  return quoti;
+}
+BigInt64 power(BigInt64 base, BigInt64 exponent) {
+  if (exponent.isNegative)
+    throw std::invalid_argument("Invalid input: negative exponent");
```

Should return an error code?

---

wrapped with try catch